File: clients/android/app/src/main/java/com/clipjot/android/data/prefs/SettingsManager.java
package com.clipjot.android.data.prefs;

import android.content.Context;
import android.content.SharedPreferences;

import com.clipjot.android.BuildConfig;

/**
 * Non-sensitive settings storage using regular SharedPreferences.
 */
public class SettingsManager {

    private static final String PREFS_NAME = "clipjot_settings";
    private static final String KEY_BACKEND_URL = "backend_url";
    private static final String KEY_PENDING_URL = "pending_bookmark_url";
    private static final String KEY_PENDING_TITLE = "pending_bookmark_title";
    private static final String KEY_USER_EMAIL = "user_email";
    private static final String KEY_PANEL_EXPANDED = "panel_expanded";

    private final SharedPreferences prefs;

    public SettingsManager(Context context) {
        prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
    }

    /**
     * Get the configured backend URL.
     */
    public String getBackendUrl() {
        return prefs.getString(KEY_BACKEND_URL, BuildConfig.DEFAULT_BACKEND_URL);
    }

    /**
     * Set the backend URL.
     */
    public void setBackendUrl(String url) {
        // Remove trailing slashes
        String normalizedUrl = url != null ? url.replaceAll("/+$", "") : BuildConfig.DEFAULT_BACKEND_URL;
        prefs.edit().putString(KEY_BACKEND_URL, normalizedUrl).apply();
    }

    /**
     * Store a pending bookmark (when user needs to login first).
     */
    public void setPendingBookmark(String url, String title) {
        prefs.edit()
                .putString(KEY_PENDING_URL, url)
                .putString(KEY_PENDING_TITLE, title)
                .apply();
    }

    /**
     * Check if there's a pending bookmark.
     */
    public boolean hasPendingBookmark() {
        return prefs.getString(KEY_PENDING_URL, null) != null;
    }

    /**
     * Get the pending bookmark URL.
     */
    public String getPendingBookmarkUrl() {
        return prefs.getString(KEY_PENDING_URL, null);
    }

    /**
     * Get the pending bookmark title.
     */
    public String getPendingBookmarkTitle() {
        return prefs.getString(KEY_PENDING_TITLE, null);
    }

    /**
     * Clear the pending bookmark.
     */
    public void clearPendingBookmark() {
        prefs.edit()
                .remove(KEY_PENDING_URL)
                .remove(KEY_PENDING_TITLE)
                .apply();
    }

    /**
     * Store the user's email (for display in settings).
     */
    public void setUserEmail(String email) {
        prefs.edit().putString(KEY_USER_EMAIL, email).apply();
    }

    /**
     * Get the stored user email.
     */
    public String getUserEmail() {
        return prefs.getString(KEY_USER_EMAIL, null);
    }

    /**
     * Clear all user data (logout).
     */
    public void clearUserData() {
        prefs.edit()
                .remove(KEY_USER_EMAIL)
                .remove(KEY_PENDING_URL)
                .remove(KEY_PENDING_TITLE)
                .apply();
    }

    /**
     * Get the panel expanded state (defaults to true).
     */
    public boolean isPanelExpanded() {
        return prefs.getBoolean(KEY_PANEL_EXPANDED, true);
    }

    /**
     * Set the panel expanded state.
     */
    public void setPanelExpanded(boolean expanded) {
        prefs.edit().putBoolean(KEY_PANEL_EXPANDED, expanded).apply();
    }

    /**
     * Reset all settings to defaults (keeps user logged in).
     */
    public void resetToDefaults() {
        prefs.edit()
                .remove(KEY_BACKEND_URL)
                .remove(KEY_PANEL_EXPANDED)
                .apply();
    }
}

File: clients/ios/ClipJot/ClipJot/Core/Storage/SettingsManager.swift
import Foundation

/// Manages app settings using UserDefaults with App Group for Share Extension access.
/// Equivalent to Android's SettingsManager with SharedPreferences.
final class SettingsManager {
    static let shared = SettingsManager()

    // App Group identifier for sharing with Share Extension
    private let appGroupIdentifier = "group.com.clipjot.shared"

    // Keys
    private enum Keys {
        static let backendUrl = "backend_url"
        static let userEmail = "user_email"
        static let quickSaveEnabled = "quick_save_enabled"
        static let formExpanded = "form_expanded"
        static let pendingBookmarkUrl = "pending_bookmark_url"
        static let pendingBookmarkTitle = "pending_bookmark_title"
    }

    // Default values
    private let defaultBackendUrl = "https://clipjot.net"

    // Use shared UserDefaults for App Group
    private var defaults: UserDefaults {
        UserDefaults(suiteName: appGroupIdentifier) ?? .standard
    }

    private init() {}

    // MARK: - Backend URL

    /// The backend server URL
    var backendUrl: String {
        get {
            let url = defaults.string(forKey: Keys.backendUrl) ?? defaultBackendUrl
            // Remove trailing slash for consistency
            return url.hasSuffix("/") ? String(url.dropLast()) : url
        }
        set {
            // Remove trailing slash before saving
            let url = newValue.hasSuffix("/") ? String(newValue.dropLast()) : newValue
            defaults.set(url, forKey: Keys.backendUrl)
        }
    }

    /// Reset backend URL to default
    func resetBackendUrl() {
        defaults.removeObject(forKey: Keys.backendUrl)
    }

    // MARK: - User Info

    /// Currently logged in user's email (for display purposes)
    var userEmail: String? {
        get { defaults.string(forKey: Keys.userEmail) }
        set { defaults.set(newValue, forKey: Keys.userEmail) }
    }

    // MARK: - Feature Toggles

    /// Quick save mode - save immediately without showing edit form
    var quickSaveEnabled: Bool {
        get { defaults.bool(forKey: Keys.quickSaveEnabled) }
        set { defaults.set(newValue, forKey: Keys.quickSaveEnabled) }
    }

    /// Whether the bookmark form is expanded (remembers last state)
    var formExpanded: Bool {
        get { defaults.bool(forKey: Keys.formExpanded) }
        set { defaults.set(newValue, forKey: Keys.formExpanded) }
    }

    // MARK: - Pending Bookmark (for Share Extension when not logged in)

    /// Store a pending bookmark when user shares but isn't logged in
    func setPendingBookmark(url: String, title: String?) {
        defaults.set(url, forKey: Keys.pendingBookmarkUrl)
        defaults.set(title, forKey: Keys.pendingBookmarkTitle)
    }

    /// Get pending bookmark if one exists
    func getPendingBookmark() -> (url: String, title: String?)? {
        guard let url = defaults.string(forKey: Keys.pendingBookmarkUrl) else {
            return nil
        }
        let title = defaults.string(forKey: Keys.pendingBookmarkTitle)
        return (url, title)
    }

    /// Check if there's a pending bookmark
    var hasPendingBookmark: Bool {
        defaults.string(forKey: Keys.pendingBookmarkUrl) != nil
    }

    /// Clear pending bookmark after it's been saved
    func clearPendingBookmark() {
        defaults.removeObject(forKey: Keys.pendingBookmarkUrl)
        defaults.removeObject(forKey: Keys.pendingBookmarkTitle)
    }

    // MARK: - Logout

    /// Clear all user-related settings (called on logout)
    func clearUserData() {
        userEmail = nil
        clearPendingBookmark()
    }

    /// Reset all settings to defaults
    func resetAll() {
        let keys = [
            Keys.backendUrl,
            Keys.userEmail,
            Keys.quickSaveEnabled,
            Keys.formExpanded,
            Keys.pendingBookmarkUrl,
            Keys.pendingBookmarkTitle,
        ]
        keys.forEach { defaults.removeObject(forKey: $0) }
    }
}
